resources/schema.py

```diff
@@ -328,6 +340,52 @@ def mutate_and_get_payload(cls, root, info, **kwargs):
         return DeleteBerthTypeMutation()
 
 
+class CreateHarborMutation(graphene.ClientIDMutation):
+    class Input(AbstractAreaInput):
+        municipality = graphene.String()
+        image_file = graphene.String()
+        availability_level = graphene.ID()
+        number_of_places = graphene.Int()
+        maximum_width = graphene.Int()
+        maximum_length = graphene.Int()
+        maximum_depth = graphene.Int()
+        name = graphene.String()
+        street_address = graphene.String()
+
+    harbor = graphene.Field(HarborNode)
+
+    @classmethod
+    @login_required
+    @superuser_required
+    @transaction.atomic
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        # TODO: Should check if the user has permissions to
+        # delete the specific resource
+        lang = get_language()
+
+        availability_level_id = kwargs.pop("availability_level", None)
```

Comment:
actually, would it be okay to just pass `availability_level_id` to `Harbor.objects.language(lang).create(**params)`? would it not assign proper availability level or `null`, when it was not given? 🤔 would save us a few lines here.

Reply:
kinda like what you do for `municipality`

Reply:
The reason why I did it like that is to check if the AL exists when it's passed.

https://github.com/City-of-Helsinki/berth-reservations/pull/115/files#diff-09f1721fc193a5a442fee1cf24f5bd49R370
```python
            try:
                availability_level = AvailabilityLevel.objects.language(lang).get(
                    pk=availability_level_id
                )
            except AvailabilityLevel.DoesNotExist as e:
                raise VenepaikkaGraphQLError(e)
```

For `municipality` I'm assuming it will work bc the model comes from an external dependency (`munigeo`), but I could also check like that.

Reply:
it does come from `munigeo`, but nothing guarantees that the database has populated data or that the user passed appropriate municipality identifier here 🤔 

I think `Harbor.objects.create(..., availability_level_id=<non-existend-id>, ...)` would raise an error (and no harbor will be created either), which would be an okay result for this case. Could you check this behaviour locally?

Reply:
Sure, I'll confirm if it works. 👍

Reply:
It returns the following error, so it works, maybe I'll wrap it on a `VenepaikkaGraphQLError`

```json
{
  "errors": [
    {
      "message": "insert or update on table \"resources_harbor\" violates foreign key constraint \"resources_harbor_availability_level_i_5d49f83c_fk_resources\"\nDETAIL:  Key (availability_level_id)=(110) is not present in table \"resources_availabilitylevel\".\n",
      "locations": [
        {
          "line": 59,
          "column": 3
        }
      ],
      "path": [
        "createHarbor"
      ]
    }
  ],
  "data": {
    "createHarbor": null
  }
}
```

Reply:
yeah, that is what we expected pretty much. hmmm, 🤔 but it would be `IntegrityError` and we would not be able to tell the frontend, which FK caused it. 🤔

Reply:
I'm trying to figure out what actually comes out of it, and `.create()` doesn't raise an error, the error is raised until you try to access the `availability_level`

This:
```python
try:
  print("foo")
  print(availability_level_id)
  harbor = Harbor.objects.language(lang).create(**kwargs)
  print(harbor.avaliability_level)
  print("bar")
except Exception as e:
  print("bam")
  raise VenepaikkaGraphQLError(e)
  print("baz")
```

raises 
``` "'Harbor' object has no attribute 'avaliability_level'"```

So, if we want to have more control, we should check it in advance.

Reply:
the time has come to `git fetch && git checkout harbor-mutations` and get dirty with it 😸

Reply:
![GS](https://media.giphy.com/media/sDjIG2QtbXKta/giphy.gif)

Reply:
ok, after some investigation, it turned out that `transaction.atomic` would eat any `DatabaseError` that could get raised during the `INSERT` (from [here and onwards](https://github.com/django/django/blob/master/django/db/transaction.py#L217)) and that is why this code would never end up in the `except` clause:
```python
try:
    harbor = Harbor.objects.language(lang).create(**kwargs)
except Exception as e:
    raise VenepaikkaGraphQLError(e)
```

so, we end up with two options:
1. try to `.get()` `AvailabilityLevel` and `Municipality` based on the IDs from the user input, but I think that is not really that necessary (at least not for the current business logic) and costs some extra hits to the DB.
2. just pass the entire input to `.create()` and let `atomic` catch and handle any exceptions. This way we do the entire business logic in just one DB hit and with fewer lines of code. We could even rename some of the input fields and it will be super succinct:
```python
class CreateHarborMutation(graphene.ClientIDMutation):
    class Input(AbstractAreaInput):
        municipality_id = graphene.String()
        image_file = graphene.String()
        availability_level_id = graphene.ID()
        number_of_places = graphene.Int()
        maximum_width = graphene.Int()
        maximum_length = graphene.Int()
        maximum_depth = graphene.Int()
        name = graphene.String()
        street_address = graphene.String()

    harbor = graphene.Field(HarborNode)

    @classmethod
    @login_required
    @superuser_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        # TODO: Should check if the user has permissions to
        # delete the specific resource
        lang = get_language()

        harbor = Harbor.objects.language(lang).create(**kwargs)
        return CreateHarborMutation(harbor=harbor)
```
the error message that the frontend would get from our backend is not that pretty, as it will be simply the `DatabaseError` casted to a string, but I guess they will be more or less ok with it 😅 
`open-city-profile` kinda does similar thing, when it calls [this helper function](https://github.com/City-of-Helsinki/open-city-profile/blob/develop/profiles/utils.py#L15) from [this mutation](https://github.com/City-of-Helsinki/open-city-profile/blob/develop/profiles/schema.py#L331)

what do you think of the simplistic option number 2?

Reply:
Option 3 would be to do something like [this](https://docs.djangoproject.com/en/3.0/topics/db/transactions/#controlling-transactions-explicitly)

```python
@classmethod
@login_required
@superuser_required
@transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        # TODO: Should check if the user has permissions to
        # delete the specific resource
        lang = get_language()

        truy @transaction.atomic
        harbor = Harbor.objects.language(lang).create(**kwargs)
        return CreateHarborMutation(harbor=harbor)
```

It would take a bit more effort adding that on every mutation, but would give us finer control to handle errors.

Reply:
Option 3 would be to do something like [this](https://docs.djangoproject.com/en/3.0/topics/db/transactions/#controlling-transactions-explicitly)

```python
@classmethod
@login_required
@superuser_required
@transaction.atomic
def mutate_and_get_payload(cls, root, info, **kwargs):
  lang = get_language()

  try:
    with transaction.atomic():
      harbor = Harbor.objects.language(lang).create(**kwargs)
  except IntegrityError as e:
    raise VenepaikkaGraphQLError(e)
  return CreateHarborMutation(harbor=harbor)
```

It would take a bit more effort adding that on every mutation, but would give us finer control to handle errors.